Fix multiple overlapping bootstraps; improve replay time and startup behaviours

belliottsmith opened the pull request:
…ure the DurableBefore Majority condition holds transitively,

since later quorums may include the bootstrapping node which does not participate in the durability of preceding transactions Also Improve:
 - Apply MaxDecidedRX filtering to CommandsForKey RX dependencies
 - optimise AbstractRanges.sliceMinimal (burn test hotspot)
 - Don't start shard schedulers on topology changes if not already started
 - Only registerTransitive if waitingOnSync
 - Reserve DurabilityRequest until ShardDurability is started
 - Don't notify progress log if stopped

belliottsmith committed:
Bootstrap must first wait for a quorum to apply the sync point to ensure the DurableBefore Majority condition holds transitively,
since later quorums may include the bootstrapping node which does not participate in the durability of preceding transactions
Also Improve:
 - Apply MaxDecidedRX filtering to CommandsForKey RX dependencies
 - optimise AbstractRanges.sliceMinimal (burn test hotspot)
 - Don't start shard schedulers on topology changes if not already started
 - Only registerTransitive if waitingOnSync
 - Reserve DurabilityRequest until ShardDurability is started
 - Don't notify progress log if stopped

belliottsmith committed:
- use normal machinery for register transitive to ensure durability

ifesdjeen commented on accord-core/src/main/java/accord/local/durability/ShardDurability.java:
```diff
@@ -346,6 +346,9 @@ synchronized void retry(Ranges ranges)
 
         synchronized void start()
         {
+            if (!isStarted())
```
Just for readability, could you maybe make it `if (!ShardDurability.this.isStarted())` ? Probably superfluous, but I think it might be useful to disambiguate.

ifesdjeen commented on accord-core/src/main/java/accord/local/Bootstrap.java:
```diff
@@ -135,20 +138,24 @@ void start(SafeCommandStore safeStore)
+                .sync("Bootstrap " + commitRanges + " for " + safeStore.commandStore(), globalSyncId, commitRanges, NoLocal, Quorum, 1L, TimeUnit.HOURS)
+                .flatMap(success -> commandStore.build((PreLoadContext.Empty) () -> "Start Bootstrap RX", safeStore0 -> {
+                    // we submit a separate execution so that we know markBootstrapping is durable before we initiate the fetch
+                    store.markBootstrapping(safeStore0, globalSyncId, commitRanges);
```
Do we maybe want to mark bootstrapping before we wait for durability majority? This used to be the first step we were making previously if I understand the left hand side correctly.

belliottsmith replied:
The reason I haven't, and I should add a comment to this effect, is that I don't want us to participate in any quorum until after a majority quorum has been reached (so that we aren't counted towards the DurableBefore Majority condition, as we don't really count for anything pre-bootstrap, as we don't wait for it). 

There is certainly a cleaner and more robust way to do this, where we track this explicitly and ensure this information is incorporated when counting votes. But, this suffices for now.

belliottsmith replied:
To be fair, it might be a relatively small thing to add that more robust logic as part of this patch. I'll mull on it overnight.

ifesdjeen commented on accord-core/src/main/java/accord/local/CommandSummaries.java:
```diff
@@ -60,185 +62,279 @@ enum SummaryStatus
+        protected final Kinds testKind;
+        protected final TxnId primaryTxnId, findAsDep, minTxnId, minDecidedId;
+        protected final Timestamp maxTxnId;
+//        protected final TxnId minDecidedId;
```
nit: minDecidedId already added as a field above, can be cleaned up here

ifesdjeen commented on accord-core/src/main/java/accord/local/CommandStore.java:
```diff
@@ -666,11 +666,33 @@ protected void updatedRedundantBefore(SafeCommandStore safeStore, RedundantBefor
+                return true;
+        }
+
+        return true;
```
Should this return false maybe, since we didn't get any intersection matches that would return true?

belliottsmith replied:
you're right, though this would have been fine (just caused some unnecessary work). I've actually tweaked this a little in the follow-up that addresses potential issues with `registerTransitive`.

ifesdjeen commented on accord-core/src/main/java/accord/local/CommandSummaries.java:
```diff
@@ -60,185 +62,279 @@ enum SummaryStatus
+        public IsDep isDep()
         {
-            return new Summary(txnId, executeAt, status, participants.slice(slice, Minimal), dep, findAsDep);
+            int ordinal = encoded >> IS_DEP_SHIFT;
```
should we maybe first check for equality with min value and then bit shift?

belliottsmith replied:
hrm. I didn't realise I had changed this in this patch. something must have gone wrong with a rebase. I don't quite follow what you mean, but let me check what happened in some rebase before we figure that out.

belliottsmith replied:
Oh, I remember. This was intertwined with the changes to CFK filtering.

belliottsmith replied:
Ok, I get it. Have pushed a change.

ifesdjeen commented on accord-core/src/main/java/accord/local/CommandSummaries.java:
```diff
@@ -60,185 +62,279 @@ enum SummaryStatus
+                return false;
 
-            private boolean isEligibleDep(SummaryStatus status, TxnId findAsDep, TxnId txnId, Timestamp executeAt)
+            // NOTE: we CANNOT safely filter on first element, as we may have pruned dependencies we need to witness
```
It is not clear to me here what you mean by first element; also because we then filter by _last_

belliottsmith replied:
I mean we are filtering based on the last element in the CFK, and while it might seem like we could do equivalent filtering based on the first/min element and maxTxnId, we can't. I will tweak the wording.

ifesdjeen commented on accord-core/src/main/java/accord/local/CommandSummaries.java:
```diff
@@ -60,185 +62,279 @@ enum SummaryStatus
-                if (intersecting.isEmpty())
+            if (mayFilterAsDecided)
+            {
+                TxnId minDecidedId = minDecidedDependencyId(maxDecidedRX, intersecting, primaryTxnId);
```
Variable shadowing looks intentional in both cases, but maybe rename to include "interesting" or something similar?

belliottsmith committed:
feedback

belliottsmith committed:
feedback

ifesdjeen commented on accord-core/src/main/java/accord/local/CommandStore.java:
```diff
@@ -666,7 +663,62 @@ protected void updatedRedundantBefore(SafeCommandStore safeStore, RedundantBefor
+                break;
+
+            Ranges remaining = e.getValue().waitingOn.without(ranges);
+            if (e.getValue().waitingOn != remaining)
```
nit: we can just always assign: if it's equal, it will remain unchanged, and new value will get assigned otherwise

belliottsmith committed:
- Remove duplicate unmanaged filtering
 - Fix edge case in notification across bootstrap boundary by resubmitting Unmanaged to be recomputed
 - Fix serialization of CommandsForKey.TxnInfo.missing() collection when non-identity flag bits are present

ifesdjeen commented on accord-core/src/main/java/accord/impl/InMemoryCommandStore.java:
```diff
@@ -420,30 +429,32 @@ protected final InMemorySafeStore createSafeStore(PreLoadContext context, Ranges
+            else
+            {
+                CommandSummaries.SummaryLoader loader = CommandSummaries.SummaryLoader.loader(unsafeGetRedundantBefore(), unsafeGetMaxDecidedRX(), context);
+                for (GlobalCommandsForKey global : this.commandsForKey.values())
```
There should be a way to avoid scanning over all keys linearly here, maybe with a range multi-map or something? Not saying that we should do this in this patch, but maybe worth adding a TODO?

belliottsmith replied:
This doesn't really need to be a range multi map, we can already search it just fine if we want to by iterating over the keys and then slicing to a subMap - we do it elsewhere, just wasn't super important here. Happy to do it now if you like.

ifesdjeen commented on accord-core/src/main/java/accord/local/cfk/Serialize.java:
```diff
@@ -77,6 +77,7 @@ public class Serialize
     private static final int HAS_BALLOT_HEADER_BIT_SHIFT = 2;
     private static final int HAS_STATUS_OVERRIDES_HEADER_BIT = 0x8;
     private static final int HAS_STATUS_OVERRIDES_HEADER_BIT_SHIFT = 3;
+    private static final int HAS_MISSING_DEPS_FLAGS_HEADER_BIT = 0x10;
```
nit: maybe add something like "bit 4 is set if there are any queries with missing deps flags" into preamble header description

ifesdjeen commented on accord-core/src/main/java/accord/local/CommandStore.java:
```diff
@@ -666,7 +663,62 @@ protected void updatedRedundantBefore(SafeCommandStore safeStore, RedundantBefor
+                break;
+
+            Ranges unmark = e.getValue().waitingOnDurable.slice(ranges, Minimal);
+            if (!unmark.isEmpty())
```
should we disassociate the key if empty?

belliottsmith replied:
I don't follow. Do you mean remove the key from the collection?

ifesdjeen commented on accord-core/src/main/java/accord/local/CommandSummaries.java:
```diff
@@ -60,185 +62,279 @@ enum SummaryStatus
+        {
+            Invariants.require(primaryTxnId != null);
+            TxnId minTxnId = redundantBefore.min(keysOrRanges, Bounds::gcBefore);
+            Timestamp maxTxnId = loadKeysFor == RECOVERY || !primaryTxnId.is(ExclusiveSyncPoint) ? Timestamp.MAX : primaryTxnId;
```
nit: maybe best remove `!` and flip arguments?

ifesdjeen replied:
or perhaps:

```
       185 +              Timestamp maxTxnId;
       186 +              TxnId findAsDep;
       187 +              Kinds kinds;
       188 +              if (loadKeysFor == RECOVERY)
       189 +              {
       190 +                  maxTxnId = Timestamp.MAX;
       191 +                  findAsDep = primaryTxnId;
       192 +                  kinds = primaryTxnId.witnesses().or(primaryTxnId.witnessedBy());
       193 +              }
       194 +              else
       195 +              {
       196 +                  maxTxnId = primaryTxnId.is(ExclusiveSyncPoint) ? primaryTxnId : Timestamp.MAX;
       197 +                   findAsDep = null;
       198 +                  kinds = primaryTxnId.witnesses().or(Nothing);
       199 +              }

```

belliottsmith committed:
- Include only the first higher RX for each key on recovery
